drivers/wifi/eswifi/eswifi_socket.c

```diff
@@ -0,0 +1,293 @@
+		goto done;
+	}
+
+	LOG_DBG("payload sz = %d", len);
+
+	pkt = net_pkt_rx_alloc_with_buffer(eswifi->iface, len,
+					   AF_UNSPEC, 0, K_NO_WAIT);
+	if (!pkt) {
+		LOG_ERR("Cannot allocate rx packet");
+		goto done;
+	}
+
+	if (!net_pkt_write(pkt, data, len)) {
+		LOG_WRN("Incomplete buffer copy");
+	}
+
+	eswifi_unlock(eswifi);
+
+	net_pkt_cursor_init(pkt);
+	socket->recv_cb(socket->context, pkt,
+			NULL, NULL, 0, socket->recv_data);
+
+	eswifi_lock(eswifi);
+
+	k_sem_give(&socket->read_sem);
+	k_yield();
+
+done:
+	err = k_delayed_work_submit_to_queue(&eswifi->work_q,
+					     &socket->read_work,
+					     500);
+	if (err) {
+		LOG_ERR("Rescheduling socket read error");
+	}
+
+	eswifi_unlock(eswifi);
+}
+
+int __eswifi_off_start_client(struct eswifi_dev *eswifi,
+			      struct eswifi_off_socket *socket)
+{
+	struct sockaddr *addr = &socket->peer_addr;
+	struct in_addr *sin_addr = &net_sin(addr)->sin_addr;
+	int err;
+
+	LOG_DBG("");
+
+	__select_socket(eswifi, socket->index);
+
+	/* Set Remote IP */
+	snprintf(eswifi->buf, sizeof(eswifi->buf), "P3=%u.%u.%u.%u\r",
+		 sin_addr->s4_addr[0], sin_addr->s4_addr[1],
+		 sin_addr->s4_addr[2], sin_addr->s4_addr[3]);
+
+	err = eswifi_at_cmd(eswifi, eswifi->buf);
+	if (err < 0) {
+		LOG_ERR("Unable to set remote ip");
+		return -EIO;
+	}
+
+	/* Set Remote Port */
+	snprintf(eswifi->buf, sizeof(eswifi->buf), "P4=%d\r",
+		(u16_t)sys_be16_to_cpu(net_sin(addr)->sin_port));
+	err = eswifi_at_cmd(eswifi, eswifi->buf);
+	if (err < 0) {
+		LOG_ERR("Unable to set remote port");
+		return -EIO;
+	}
+
+	/* Start TCP/UDP client */
+	snprintf(eswifi->buf, sizeof(eswifi->buf), "P6=1\r");
+	err = eswifi_at_cmd(eswifi, eswifi->buf);
+	if (err < 0) {
+		LOG_ERR("Unable to start TCP/UDP client");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+int __eswifi_accept(struct eswifi_dev *eswifi, struct eswifi_off_socket *socket)
+{
+	char cmd[] = "P5=1\r";
+
+	if (socket->state != ESWIFI_SOCKET_STATE_NONE) {
+		/* we can only handle one connection at a time */
+		return -EBUSY;
+	}
+
+	__select_socket(eswifi, socket->index);
+
+	/* Start TCP Server */
+	if (eswifi_at_cmd(eswifi, cmd) < 0) {
+		LOG_ERR("Unable to start TCP server");
+		return -EIO;
+	}
+
+	LOG_DBG("TCP Server started");
+	socket->state = ESWIFI_SOCKET_STATE_ACCEPTING;
+
+	return 0;
+}
+
+int __eswifi_socket_free(struct eswifi_dev *eswifi,
+			 struct eswifi_off_socket *socket)
+{
+	if ((socket->type == ESWIFI_TRANSPORT_TCP ||
+	     socket->type == ESWIFI_TRANSPORT_TCP_SSL) &&
+	     socket->state != ESWIFI_SOCKET_STATE_CONNECTED)
+		return -ENOTCONN;
+
+	__select_socket(eswifi, socket->index);
+	k_delayed_work_cancel(&socket->read_work);
+
+	__select_socket(eswifi, socket->index);
+	__stop_socket(eswifi, socket);
+
+	return 0;
+}
+
+int __eswifi_socket_new(struct eswifi_dev *eswifi, int family, int type,
+			int proto, void *context)
+{
+	struct eswifi_off_socket *socket = NULL;
+	int err, i;
+
+	LOG_DBG("");
+
+	if (family != AF_INET) {
```

Comment:
BTW, do you have somewhere a check whether IPv4 is enabled at all?

Reply:
Yes it is handled in eswifi_core.c